File: 2nd Year/2nd Semester/MPP/PROIECT-JAVA/Laborator/TriatlonPersistence/src/main/java/triatlon/persistence/ArbitruRepository.java
package triatlon.persistence;

import triatlon.model.Arbitru;
import triatlon.persistence.JdbcUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ArbitruRepository implements IArbitruRepository {

    private final JdbcUtils jdbcUtils;
    private static final Logger logger = LogManager.getLogger();

    public ArbitruRepository(Properties properties) {
        logger.info("Initializing ArbitruRepository with properties: {}", properties);
        jdbcUtils = new JdbcUtils(properties);
    }

    @Override
    public Arbitru findOne(Integer integer) {
        logger.traceEntry("finding Arbitru with id: {}", integer);
        String sql = "SELECT * FROM Arbitru WHERE id = ?";
        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setInt(1, integer);
            try (ResultSet rs = preparedStatement.executeQuery()) {
                if (rs.next()) {
                    return new Arbitru(
                            rs.getInt("id"),
                            rs.getString("username"),
                            rs.getString("password"),
                            rs.getString("first_name"),
                            rs.getString("last_name")
                    );
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        return null;
    }

    @Override
    public Iterable<Arbitru> findAll() {
        List<Arbitru> arbitrii = new ArrayList<>();
        logger.traceEntry("finding all Arbitrii");
        String sql = "SELECT * FROM Arbitru";
        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql);
             ResultSet resultSet = preparedStatement.executeQuery()) {
            while (resultSet.next()) {
                Arbitru arbitru = new Arbitru(
                        resultSet.getInt("id"),
                        resultSet.getString("username"),
                        resultSet.getString("password"),
                        resultSet.getString("first_name"),
                        resultSet.getString("last_name")
                );
                arbitrii.add(arbitru);
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
        return arbitrii;
    }

    @Override
    public void save(Arbitru entity) {
        logger.traceEntry("saving Arbitru {}", entity);
        String sql = "insert into Arbitru (username, password, first_name, last_name) values (?, ?, ?, ?)";
        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql, PreparedStatement.RETURN_GENERATED_KEYS)) {
            preparedStatement.setString(1, entity.getUsername());
            preparedStatement.setString(2, entity.getPassword());
            preparedStatement.setString(3, entity.getFirst_name());
            preparedStatement.setString(4, entity.getLast_name());
            int affectedRows = preparedStatement.executeUpdate();
            if (affectedRows > 0) {
                try (ResultSet generatedKeys = preparedStatement.getGeneratedKeys()) {
                    if (generatedKeys.next()) {
                        entity.setId(generatedKeys.getInt(1));
                    }
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
    }

    @Override
    public void delete(Integer integer) {
        logger.traceEntry("deleting Arbitru with id: {}", integer);
        String sql = "DELETE FROM Arbitru WHERE id = ?";
        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setInt(1, integer);
            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
    }

    @Override
    public void update(Arbitru entity) {
        logger.traceEntry("updating Arbitru {}", entity);
        String sql = "UPDATE Arbitru SET username = ?, password = ?, first_name = ?, last_name = ? WHERE id = ?";
        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setString(1, entity.getUsername());
            preparedStatement.setString(2, entity.getPassword());
            preparedStatement.setString(3, entity.getFirst_name());
            preparedStatement.setString(4, entity.getLast_name());
            preparedStatement.setInt(5, entity.getId());
            preparedStatement.executeUpdate();  // Added missing executeUpdate call
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
    }

    @Override
    public Arbitru findBy(String username, String password) {
        logger.traceEntry("finding arbitru with username : {} and password: {}", username, password);
        String sql = "SELECT * FROM Arbitru WHERE username = ? AND password = ?";
        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {
            preparedStatement.setString(1, username);
            preparedStatement.setString(2, password);
            try (ResultSet rs = preparedStatement.executeQuery()) {
                if (rs.next()) {
                    return new Arbitru(
                            rs.getInt("id"),
                            rs.getString("username"),
                            rs.getString("password"),
                            rs.getString("first_name"),
                            rs.getString("last_name")
                    );
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
        return null;
    }
}

File: 2nd Year/2nd Semester/MPP/Triatlon-C#/Lab1/Repository/ArbitruRepository.cs
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Data.SQLite;
using Lab1.Domain;
using log4net;

namespace Lab1.Repository {
    public class ArbitruRepository : IArbitruRepository {
        private static readonly ILog logger = LogManager.GetLogger(typeof(ArbitruRepository));
        public readonly string connectionString;

        public ArbitruRepository() {
            connectionString = ConfigurationManager.ConnectionStrings["DefaultConnection"].ConnectionString;
        }

        public void Delete(int id) {
            logger.Debug("Enter deleting Arbitru with id: " + id);
            string sql = "DELETE FROM Arbitru WHERE id=@id";
            try {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString)) {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection)) {
                        command.Parameters.AddWithValue("@id", id);
                        command.ExecuteNonQuery();
                    }
                    connection.Close();
                }
                logger.Debug("Exit deleting Arbitru with id: " + id);
            }
            catch (SQLiteException ex) {
                logger.Error("Error deleting Arbitru with id: " + id, ex);
            }
        }

        public List<Arbitru> FindAll() {
            logger.Debug("Enter FindAll Arbitru");
            List<Arbitru> arbitri = new List<Arbitru>();
            string sql = "SELECT * FROM Arbitru";
            try {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString)) {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection)) {
                        using (SQLiteDataReader reader = command.ExecuteReader()) {
                            while (reader.Read()) {
                                int id = reader.GetInt32(0);
                                string username = reader.GetString(1);
                                string password = reader.GetString(2);
                                string firstName = reader.GetString(3);
                                string lastName = reader.GetString(4);
                                Arbitru arbitru = new Arbitru(id, username, password, firstName, lastName);
                                arbitri.Add(arbitru);
                                logger.Debug("Found Arbitru: " + arbitru);
                            }
                        }
                    }
                    connection.Close();
                }
            }
            catch (SQLiteException ex) {
                logger.Error("Error finding all Arbitru", ex);
            }
            logger.Debug("Exit FindAll Arbitru");
            return arbitri;
        }

        public Arbitru FindOne(int id) {
            logger.Debug("Enter FindOne Arbitru with id: " + id);
            Arbitru arbitru = null;
            string sql = "SELECT * FROM Arbitru WHERE id=@id";
            try {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString)) {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection)) {
                        command.Parameters.AddWithValue("@id", id);
                        using (SQLiteDataReader reader = command.ExecuteReader()) {
                            if (reader.Read()) {
                                string username = reader.GetString(1);
                                string password = reader.GetString(2);
                                string firstName = reader.GetString(3);
                                string lastName = reader.GetString(4);
                                arbitru = new Arbitru(id, username, password, firstName, lastName);
                                logger.Debug("Found Arbitru: " + arbitru);
                            }
                        }
                    }
                    connection.Close();
                }
                logger.Debug("Exit FindOne Arbitru with id: " + id);
            }
            catch (SQLiteException ex) {
                logger.Error("Error finding Arbitru with id: " + id, ex);
            }
            return arbitru;
        }

        public void Save(Arbitru entity) {
            logger.Debug("Enter Save Arbitru: " + entity);

            string sql = "INSERT INTO Arbitru (username, password, first_name, last_name) VALUES (@username, @password, @firstName, @lastName)";
            try {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString)) {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection)) {
                        command.Parameters.AddWithValue("@username", entity.Username);
                        command.Parameters.AddWithValue("@password", entity.Password);
                        command.Parameters.AddWithValue("@firstName", entity.FirstName);
                        command.Parameters.AddWithValue("@lastName", entity.LastName);
                        int affectedRows = command.ExecuteNonQuery();
                        if (affectedRows > 0) {
                            command.CommandText = "SELECT last_insert_rowid()";
                            entity.Id = Convert.ToInt32(command.ExecuteScalar());
                        }
                        logger.Debug("Saved Arbitru: " + entity);
                    }
                    connection.Close();
                }
                logger.Debug("Exit Save Arbitru: " + entity);
            }
            catch (SQLiteException ex) {
                logger.Error("Error saving Arbitru: " + entity, ex);
            }
        }



        public void Update(Arbitru entity) {
            logger.Debug("Enter Update Arbitru: " + entity);
            string sql = "UPDATE Arbitru SET username=@username, password=@password, first_name=@firstName, last_name=@lastName WHERE id=@id";
            try {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString)) {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection)) {
                        command.Parameters.AddWithValue("@id", entity.Id);
                        command.Parameters.AddWithValue("@username", entity.Username);
                        command.Parameters.AddWithValue("@password", entity.Password);
                        command.Parameters.AddWithValue("@firstName", entity.FirstName);
                        command.Parameters.AddWithValue("@lastName", entity.LastName);
                        command.ExecuteNonQuery();
                    }
                    connection.Close();
                }
                logger.Debug("Exit Update Arbitru: " + entity);
            }
            catch (SQLiteException ex) {
                logger.Error("Error updating Arbitru: " + entity, ex);
            }
        }

        public Arbitru FindBy(string username, string password)
        {
            logger.Debug("Enter FindBy Arbitru with username: " + username + " and password: " + password);
            Arbitru arbitru = null;
            string sql = "SELECT * FROM Arbitru WHERE username=@username AND password=@password";
            try {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString)) {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection)) {
                        command.Parameters.AddWithValue("@username", username);
                        command.Parameters.AddWithValue("@password", password);
                        using (SQLiteDataReader reader = command.ExecuteReader()) {
                            if (reader.Read()) {
                                int id = reader.GetInt32(0);
                                string firstName = reader.GetString(3);
                                string lastName = reader.GetString(4);
                                arbitru = new Arbitru(id, username, password, firstName, lastName);
                                logger.Debug("Found Arbitru: " + arbitru);
                            }
                        }
                    }
                    connection.Close();
                }
                logger.Debug("Exit FindBy Arbitru with username: " + username + " and password: " + password);
            }
            catch (SQLiteException ex) {
                logger.Error("Error finding Arbitru with username: " + username + " and password: " + password, ex);
            }
            return arbitru;
        }
    }
}
